**Coursebook: Exploratory Data Analysis**
- Part 2 of Data Analytics Specialization
- Course Length: 12 hours
- Last Updated: April 2019

___

- Author: [Samuel Chan](https://github.com/onlyphantom)
- Developed by [Algoritma](https://algorit.ma)'s product division and instructors team

# Background

## Top-Down Approach 

The coursebook is part of the **Data Analytics Specialization** offered by [Algoritma](https://algorit.ma). It takes a more accessible approach compared to Algoritma's core educational products, by getting participants to overcome the "how" barrier first, rather than a detailed breakdown of the "why". 

This translates to an overall easier learning curve, one where the reader is prompted to write short snippets of code in frequent intervals, before being offered an explanation on the underlying theoretical frameworks. Instead of mastering the syntactic design of the Python programming language, then moving into data structures, and then the `pandas` library, and then the mathematical details in an imputation algorithm, and its code implementation; we would do the opposite: Implement the imputation, then a succinct explanation of why it works and applicational considerations (what to look out for, what are assumptions it made, when _not_ to use it etc).

## Learn-by-Building

This coursebook is intended for participants who have completed the preceding courses offered in the **Data Analytics Developer** Specialization. This is the second course, **Exploratory Data Analysis**

The coursebook focuses on:
- Why and What: Exploratory Data Analysis
- Treating Duplicates and Missing Values 
- Table, Cross Tabulation and Pivot Table
- Date Time objects
- Categorical Data Types  

At the end of this course is a Graded Asssignment section, where you are expected to apply all that you've learned on a new dataset, and attempt the given questions.


# Data Preparation and Exploration

About 60 years ago, John Tukey defined data analysis as the "procedures for analyzing data, techniques for interpreting the results of such procedures ... and all the machinery of mathematical statistics which apply to analyzing data". His championing of EDA encouraged the development of statsitical computing packages, especially S at Bell Labs (which later inspired R).

He wrote a book titled _Exploratory Data Analysis_ arguing that too much emphasis in statistics was placed on hypothesis testing (confirmatory data analysis) while not enough was placed on the discovery of the unexpected. 

> Exploratory data analysis isolates patterns and features of the data and reveals these forcefully to the analyst.

This course aims to present a selection of EDA techniques -- some developed by John Tukey himself -- but with a special emphasis on its application to modern business analytics.

In [3]:
import pandas as pd
import numpy as np
print(pd.__version__)

0.24.2


In the previous course, we've got our hands on a few common techniques:

- `.head()` and `.tail()`
- `.describe()`
- `.shape` and `.size`
- `.axes`
- `.dtypes`

Let’s now take a look at another real life dataset and apply our newly acquired knowledge. We'll start by reading our dataset in;

In [2]:
review = pd.read_csv("data_input/clothing_reviews.csv")

review.shape

(23486, 9)

`clothing_review.csv` is a women’s clothing e-commerce dataset revolving around the reviews written by customers. This dataset includes 23,486 rows and 9 feature variables. Each row corresponds to a customer review, and includes the variables:

- `clothing_id`: refers to the specific piece being reviewed.
- `age_group`: categorical variable of the reviewers age group.
- `review_text`: string variable for the title of the review.
- `rating`: integer variable for the product score granted by the customer from 1 Worst, to 5 Best.
- `recommended_ind`: binary variable stating where the customer recommends the product where 1 is recommended, 0 is not recommended.
- `positive_feedback_count`: positive integer documenting the number of other customers who found this review positive.
- `division_name`: categorical name of the product high level division.
- `category`: categorical name of the product category.
- `review_date`: date of the review added.

Now, try to create a new cell below and peek at the first few rows of the data!

In [3]:
## Your code below


## -- Solution code
review.head()

clothing_id age_group                                        review_text  \
0          767  25 to 34  Absolutely wonderful - silky and sexy and comf...   
1         1080  25 to 34  Love this dress!  it's sooo pretty.  i happene...   
2         1077  55 to 64  I had such high hopes for this dress and reall...   
3         1049  35 to 54  I love, love, love this jumpsuit. it's fun, fl...   
4          847  35 to 54  This shirt is very flattering to all due to th...   

   rating  recommended_ind  positive_feedback_count   division_name  category  \
0       4                1                        0       Initmates  Intimate   
1       5                1                        4         General   Dresses   
2       3                0                        0         General   Dresses   
3       5                1                        0  General Petite   Bottoms   
4       5                1                        6         General      Tops   

        review_date  
0  2017-03-30 06:00  
1  2017-03-10 22:00  
2  2017-03-04 10:00  
3  2017-03-03 14:00  
4  2017-12-14 06:00

In the following chapters, we'll expand our EDA toolset with the following additions:  

- Tables
- Cross-Tables and Aggregates
- Using `aggfunc` for aggregate functions
- Pivot Tables

## Tables

One of the simplest EDA toolkit is the frequency table (contingency tables) and cross-tabulation tables. It is highly familiar, convenient, and practical for a wide array of statistical tasks. The simplest form of a table is to display counts of a `categorical` column.

In `pandas`, each column of a `DataFrame` is a `Series`. To get the counts of each unique levels in a categorical column, we can use `.value_counts()`. The resulting object is a `Series` and in descending order so that the most frequent element is on top. 

Suppose we want to know the amount of reviews obtained by each clothing `category`. Now, try and perform `.value_counts()` on the `category` column, adding either:

- `sort=False` as a parameter to prevent any sorting of elements, or
- `ascending=True` as a parameter to sort in ascending order instead

In [4]:
review.category.value_counts()

Tops        10468
Dresses      6319
Bottoms      3799
Intimate     1735
Jackets      1032
Trend         119
Name: category, dtype: int64

In [5]:
## Your code below


## -- Solution code

`crosstab` is a very versatile solution to producing frequency tables on a `DataFrame` object. Its utility really goes further than that but we'll start with a simple use-case.

Consider the following code: we use `pd.crosstab()` passing in the values to group by in the rows (`index`) and columns (`columns`) respectively. 

In [6]:
pd.crosstab(index = review['category'],columns = "count")

col_0     count
category       
Bottoms    3799
Dresses    6319
Intimate   1735
Jackets    1032
Tops      10468
Trend       119

Realize that in the code above, we're setting the row (index) to be `categories` and the function will by default compute a frequency table. 

In [7]:
pd.crosstab(index = review['category'], columns = "count", normalize = 'columns')

col_0        count
category          
Bottoms   0.161852
Dresses   0.269214
Intimate  0.073918
Jackets   0.043967
Tops      0.445978
Trend     0.005070

In the cell above, we set the values to be normalized over each columns, and this will divide each values in place over the sum of all values. This is equivalent to a manual calculation:

In [8]:
cat = pd.crosstab(index = review['category'], columns = 'counts')
cat/cat.sum()

col_0       counts
category          
Bottoms   0.161852
Dresses   0.269214
Intimate  0.073918
Jackets   0.043967
Tops      0.445978
Trend     0.005070

We can also use the same `crosstab` method to compute a cross-tabulation of two factors. Say, for example, we want to know how our customer's age group corresponds to the reviews in each clothing category. 

So, in the following cell, we use `crosstab` with the `index` references the `category` column while the `columns` references the `age_group`:

In [9]:
pd.crosstab(index = review['category'], columns = review['age_group'])

age_group  18 to 24  25 to 34  35 to 54  55 to 64  65+
category                                              
Bottoms         130       815      2156       514  184
Dresses         273      1524      3464       769  289
Intimate         80       483       914       161   97
Jackets          42       237       515       157   81
Tops            360      2098      5747      1536  727
Trend             6        15        75        15    8

This is intuitive in a way: We use `crosstab()` which, we recall, computes the count and we pass in `index` and `columns` which correspond to the row and column respectively.

When we add `margins=True` to our method call, then an extra row and column of margins (subtotals) will be included in the output:

In [10]:
pd.crosstab(index = review['category'], 
            columns = review['age_group'], 
            margins= True)

age_group  18 to 24  25 to 34  35 to 54  55 to 64   65+    All
category                                                      
Bottoms         130       815      2156       514   184   3799
Dresses         273      1524      3464       769   289   6319
Intimate         80       483       914       161    97   1735
Jackets          42       237       515       157    81   1032
Tops            360      2098      5747      1536   727  10468
Trend             6        15        75        15     8    119
All             891      5172     12871      3152  1386  23472

Say, for example, you want to analyze how the `rating` varies within each `category`. In the following cell, use `pd.crosstab()` with `rating` as the row and `category` as the column. Set `margins=True` to get a total across the row and columns. 

In [11]:
## Your code below


## -- Solution code
pd.crosstab(index=review['rating'], 
            columns=review['category'], 
            margins=True)

category  Bottoms  Dresses  Intimate  Jackets   Tops  Trend    All
rating                                                            
1             115      228        63       49    376     11    842
2             205      461        90       60    738     11   1565
3             421      838       178       91   1324     19   2871
4             785     1395       371      201   2299     26   5077
5            2273     3397      1033      631   5731     52  13117
All          3799     6319      1735     1032  10468    119  23472

If you want an extra challenge, try and modify your code above to include a `normalize` parameter. 

`normalize` accepts a boolean value, or one of `all`, `index` or `columns`. Since we want it to normalize across each row, we will set this parameter to the value of `index`.

### Aggregation Table

In the following section, we will introduce another parameter to perform aggregation on our table. The `aggfunc` parameter when present, required the `values` parameter to be specified as well. `values` is the values to aggregate according to the factors in our index and columns:

In [12]:
pd.crosstab(index=review['category'], 
            columns='Total Positive Feedbacks', 
            values=review['positive_feedback_count'],
            aggfunc='sum')

col_0     Total Positive Feedbacks
category                          
Bottoms                       8043
Dresses                      19510
Intimate                      3275
Jackets                       2916
Tops                         25407
Trend                          401

#### Knowledge Check

Create a cross-tab using `sub_category` as the index (row) and `format` as the column. Fill the values with the median of `unit_price` across each row and column. Add a subtotal to both the row and column by setting `margins=True`.

1. On average, Sugar is cheapest at...?
2. On average, Detergent is most expensive at...?

Create a new cell for your code and answer the questions above.

In [13]:
## Your code below


## -- Solution code

Reference answer:

```
pd.crosstab(index=household['sub_category'], 
            columns=household['format'], 
            values=household['unit_price'],
            aggfunc='median', margins=True)
```

### Higher-dimensional Tables

If we need to inspect our data in higher resolution, we can create cross-tabulation using more than one factor. This allows us to yield insights on a more granular level yet have our output remain relatively compact and structured:

In [14]:
pd.crosstab(index = review['age_group'], 
            columns = [review['division_name'], review['category']], 
            values = review['rating'],
            aggfunc='mean')

division_name   General                                          \
category        Bottoms   Dresses   Jackets      Tops     Trend   
age_group                                                         
18 to 24       4.386364  4.311688  4.407407  4.394619  4.400000   
25 to 34       4.221402  4.089362  4.189542  4.107581  3.461538   
35 to 54       4.251366  4.167668  4.253918  4.101067  3.894737   
55 to 64       4.381381  4.218415  4.214286  4.292339  3.538462   
65+            4.295652  4.228070  4.270833  4.227926  4.000000   

division_name General Petite                                          \
category             Bottoms   Dresses  Intimate   Jackets      Tops   
age_group                                                              
18 to 24            4.404762  4.310924  4.250000  4.400000  4.226277   
25 to 34            4.234432  4.111301  4.108696  4.083333  4.106592   
35 to 54            4.346821  4.112551  4.226027  4.326531  4.207812   
55 to 64            4.320442  4.274834  4.619048  4.355932  4.305147   
65+                 4.507246  3.957627  4.250000  4.606061  4.408333   

division_name           Initmates  
category          Trend  Intimate  
age_group                          
18 to 24       3.000000  4.277778  
25 to 34       4.000000  4.265446  
35 to 54       3.777778  4.309896  
55 to 64       4.000000  4.278571  
65+                 NaN  4.200000

## Pivot Tables

If our data is already in a `DataFrame` format, using `pd.pivot_table` can sometimes be more convenient compared to a `pd.crosstab`. 

Fortunately, much of the parameters in a `pivot_table()` function is the same as `pd.crosstab()`. The noticable difference is the use of an additional `data` parameter, which allow us to specify the `DataFrame` that is used to construct the pivot table.

We create a `pivot_table` by passing in the following:
- `data`: our `DataFrame`
- `index`: the column to be used as rows
- `columns`: the column to be used as columns
- `values`: the values used to fill in the table
- `aggfunc`: the aggregation function

In [15]:
pd.pivot_table(
    data = review,
    index = 'age_group',
    columns = ['division_name', 'category'],
    values = 'rating',
    aggfunc = 'mean'
)

division_name   General                                          \
category        Bottoms   Dresses   Jackets      Tops     Trend   
age_group                                                         
18 to 24       4.386364  4.311688  4.407407  4.394619  4.400000   
25 to 34       4.221402  4.089362  4.189542  4.107581  3.461538   
35 to 54       4.251366  4.167668  4.253918  4.101067  3.894737   
55 to 64       4.381381  4.218415  4.214286  4.292339  3.538462   
65+            4.295652  4.228070  4.270833  4.227926  4.000000   

division_name General Petite                                          \
category             Bottoms   Dresses  Intimate   Jackets      Tops   
age_group                                                              
18 to 24            4.404762  4.310924  4.250000  4.400000  4.226277   
25 to 34            4.234432  4.111301  4.108696  4.083333  4.106592   
35 to 54            4.346821  4.112551  4.226027  4.326531  4.207812   
55 to 64            4.320442  4.274834  4.619048  4.355932  4.305147   
65+                 4.507246  3.957627  4.250000  4.606061  4.408333   

division_name           Initmates  
category          Trend  Intimate  
age_group                          
18 to 24       3.000000  4.277778  
25 to 34       4.000000  4.265446  
35 to 54       3.777778  4.309896  
55 to 64       4.000000  4.278571  
65+                 NaN  4.200000

A key difference between `crosstab` and `pivot_table` is that `crosstab` uses `len` (or `count`) as the default aggregation function while `pivot_table` using the mean. Copy the code from the cell below and make a change: use `sum` as the aggregation function instead: 

In [16]:
pd.pivot_table(
    data = review,
    index = 'category',
    columns= 'age_group',
    values = 'recommended_ind'
)

age_group  18 to 24  25 to 34  35 to 54  55 to 64       65+
category                                                   
Bottoms    0.861538  0.831902  0.848794  0.873541  0.896739
Dresses    0.864469  0.791339  0.807448  0.825748  0.806228
Intimate   0.850000  0.840580  0.853392  0.857143  0.855670
Jackets    0.904762  0.780591  0.856311  0.834395  0.839506
Tops       0.866667  0.787893  0.806160  0.850911  0.863824
Trend      1.000000  0.666667  0.720000  0.733333  0.875000

In [17]:
## Your code below


## -- Solution code

## Working with Datetime

Given the program's special emphasis on business-driven analytics, one data type of particular interest to us is the `datetime`. In the first part of this coursebook, we've seen an example of `datetime` in the section introducing data types (`member.birth`).

A large portion of data science work performed by business executives involve time series and/or dates (think about the kind of data science work done by computer vision researchers, and compare that to the work done by credit rating analysts or marketing executives and this special relationship between business and datetime data becomes apparent), so adding a level of familiarity with this format will serve you well in the long run. 

As a start, let's take a look at the data types of our `DataFrame` again:

In [18]:
review.dtypes

clothing_id                 int64
age_group                  object
review_text                object
rating                      int64
recommended_ind             int64
positive_feedback_count     int64
division_name              object
category                   object
review_date                object
dtype: object

Notice that all columns are in the right data types, except for `review_date`. The correct data type for this column would have to be a `datetime`.

To convert a column `x` to a datetime, we would use:

    `x = pd.to_datetime(x)`
    

In [19]:
review['review_date'] = pd.to_datetime(review['review_date'])
review.head()

clothing_id age_group                                        review_text  \
0          767  25 to 34  Absolutely wonderful - silky and sexy and comf...   
1         1080  25 to 34  Love this dress!  it's sooo pretty.  i happene...   
2         1077  55 to 64  I had such high hopes for this dress and reall...   
3         1049  35 to 54  I love, love, love this jumpsuit. it's fun, fl...   
4          847  35 to 54  This shirt is very flattering to all due to th...   

   rating  recommended_ind  positive_feedback_count   division_name  category  \
0       4                1                        0       Initmates  Intimate   
1       5                1                        4         General   Dresses   
2       3                0                        0         General   Dresses   
3       5                1                        0  General Petite   Bottoms   
4       5                1                        6         General      Tops   

          review_date  
0 2017-03-30 06:00:00  
1 2017-03-10 22:00:00  
2 2017-03-04 10:00:00  
3 2017-03-03 14:00:00  
4 2017-12-14 06:00:00

In fact, `pandas` has a number of machineries to work with `datetime` objects. These are convenient for when we need to extract the `month`, or `year`, or `weekday_name` from `datetime`. Some common applications in business analysis include:

- `review['review_date'].dt.month`
- `review['review_date'].dt.year`
- `review['review_date'].dt.day`
- `review['review_date'].dt.dayofweek`
- `review['review_date'].dt.hour`
- `review['review_date'].dt.weekday_name`

There are also other functions that can be helpful in certain situations. Supposed we want to transform the existing `datetime` column into values of periods we can use the `.to_period` method:

- `review['review_date'].dt.to_period('D')`
- `review['review_date'].dt.to_period('W')`
- `review['review_date'].dt.to_period('M')`
- `review['review_date'].dt.to_period('Q')`

For example, if we want to count how many reviews obtained within each quarter:

In [20]:
review['review_date'].dt.to_period('Q').value_counts()

2017Q3    8470
2017Q4    4289
2018Q2    3336
2017Q2    3326
2017Q1    2215
2018Q1    1495
2018Q3     122
2015Q4      89
2015Q2      48
2016Q2      31
2015Q3      31
2014Q3      12
2015Q1      11
2016Q1       5
2016Q3       3
2016Q4       2
2014Q4       1
Freq: Q-DEC, Name: review_date, dtype: int64

### Knowledge Check
_Est. Time required: 20 minutes_

1. In the following cell, start again by reading in the `clothing_reviews.csv` dataset 
2. Convert `review_date` to `datetime`. Use `pd.to_datetime()` for this.
3. Use `x.dt.weekday_name`, assuming `x` is a datetime object to get the day of week. Assign this to a new column in your `review` Data Frame, name it `weekday`
4. Print the first 5 rows of your data to verify that your preprocessing steps are correct

Tips: In the cell above, start from:

`review = pd.read_csv("data_input/clothing_reviews.csv")`

Inspect the first 5 rows of your data and pay close attention to the `weekday` column. 

**Bonus challenge:**
5. How many reviews obtained on each day of the week? Use `pd.crosstab(index=__, columns="count")` or `x.value_counts()`.

In [21]:
## Your code below


## -- Solution code

If you've managed the above exercises, well done! Run the following cell anyway to make sure we're at the same starting point as we go into the next chapter of working with categorical data (factors). :

In [10]:
review = pd.read_csv("data_input/clothing_reviews.csv", parse_dates=['review_date'])
review['weekday'] = review['review_date'].dt.weekday_name
review.head()
pd.crosstab(index = review['weekday'], columns = "count")

col_0      count
weekday         
Friday      3157
Monday      3015
Saturday    3528
Sunday      1569
Thursday    3468
Tuesday     5032
Wednesday   3717

## Working with Categories

The official documentation from `pandas` describe the `category` data type as a tool to "represent a categorical variable in classic R fashion".

When working with categories, it is recommended both from a business point of a view and a technical one to use `pandas` categorical data type. From a business perspective, this adds clarity to the analyst's mind about the type of data he/she is working with. This informs and guides the analysis, on questions such as which statistical methods or plot types to use.

From a technical viewpoint, the memory savings -- and in turn, computation speed as well as computational resources -- can be quite significant. Specifically, the docs remarked:

> The memory usage of a `Categorical` is proportional to the number of categories plus the length of the data. In contrast, an `object` dtype is a constant times the length of the data

In [25]:
review.dtypes

clothing_id                         int64
age_group                          object
review_text                        object
rating                              int64
recommended_ind                     int64
positive_feedback_count             int64
division_name                      object
category                           object
review_date                datetime64[ns]
weekday                            object
dtype: object

From the output of `dtypes`, we see that there are three variables currently stored as `object` type where a `category` is more appropriate. This is a common diagnostic step, and one that you will employ in almost every data analysis project. 

We'll convert the `weekday` column to a categorical type using `.astype()`. `astype('int64')` converts a Series to an integer type, and `.astype(category)` logically, converts a Series to a categorical.

By default, `.astype()` will raise an error if the conversion is not successful (we call them "exceptions"). In an analysis-driven environment, this is what we usually prefer. However, in certain production settings, you don't want the exception to be raised and rather return the original object (`errors='ignore'`).

In [36]:
review['weekday'] = review['weekday'].astype('category', errors = 'raise')
review.dtypes

clothing_id                         int64
age_group                          object
review_text                        object
rating                              int64
recommended_ind                     int64
positive_feedback_count             int64
division_name                      object
category                         category
review_date                datetime64[ns]
weekday                          category
dtype: object

Go ahead and perform the other conversions in the following cell. When you're done, use `dtypes` to check that you have the categorical columns stored as `category`.

### Alternative Solutions (optional)

#### Alternative 1:

In [38]:
review.select_dtypes(exclude = 'object').head()

clothing_id  rating  recommended_ind  positive_feedback_count  category  \
0          767       4                1                        0  Intimate   
1         1080       5                1                        4   Dresses   
2         1077       3                0                        0   Dresses   
3         1049       5                1                        0   Bottoms   
4          847       5                1                        6      Tops   

          review_date   weekday  
0 2017-03-30 06:00:00  Thursday  
1 2017-03-10 22:00:00    Friday  
2 2017-03-04 10:00:00  Saturday  
3 2017-03-03 14:00:00    Friday  
4 2017-12-14 06:00:00  Thursday

In [39]:
pd.concat([
    review.select_dtypes(exclude='object'),
    review.select_dtypes(include='object').apply(
        pd.Series.astype, dtype='category'
    )
], axis=1).dtypes

clothing_id                         int64
rating                              int64
recommended_ind                     int64
positive_feedback_count             int64
category                         category
review_date                datetime64[ns]
weekday                          category
age_group                        category
review_text                      category
division_name                    category
dtype: object

#### Alternative 2:

In [40]:
objectcols = review.select_dtypes(include='object')

review[objectcols.columns] = objectcols.apply(lambda x: x.astype('category'))

review.head()

clothing_id age_group                                        review_text  \
0          767  25 to 34  Absolutely wonderful - silky and sexy and comf...   
1         1080  25 to 34  Love this dress!  it's sooo pretty.  i happene...   
2         1077  55 to 64  I had such high hopes for this dress and reall...   
3         1049  35 to 54  I love, love, love this jumpsuit. it's fun, fl...   
4          847  35 to 54  This shirt is very flattering to all due to th...   

   rating  recommended_ind  positive_feedback_count   division_name  category  \
0       4                1                        0       Initmates  Intimate   
1       5                1                        4         General   Dresses   
2       3                0                        0         General   Dresses   
3       5                1                        0  General Petite   Bottoms   
4       5                1                        6         General      Tops   

          review_date   weekday  
0 2017-03-30 06:00:00  Thursday  
1 2017-03-10 22:00:00    Friday  
2 2017-03-04 10:00:00  Saturday  
3 2017-03-03 14:00:00    Friday  
4 2017-12-14 06:00:00  Thursday

In [41]:
review.dtypes

clothing_id                         int64
age_group                        category
review_text                      category
rating                              int64
recommended_ind                     int64
positive_feedback_count             int64
division_name                    category
category                         category
review_date                datetime64[ns]
weekday                          category
dtype: object

## Missing Values and Duplicates

In the following cell, I'm going to  use `reindex` to "inject" some rows where values don't exist (clothing id 1100 through 1105) and also set `math.nan` on one of the values for `weekday`. Notice from the output that between row 3 to 8 there are at least a few rows with missing data.

In [54]:
import math
x=[i for i in range(1100, 1105)]
x.insert(2,820)
x

[1100, 1101, 820, 1102, 1103, 1104]

In [116]:
import math
x=[i for i in range(2400, 2404)]
x.insert(2,910)
x

review2 = review.head(5).copy()
review2 = review2.reindex(x)
review2 = pd.concat([review2, review.head(9)])
review2.loc[1049, "review_date"] = math.nan
review2.iloc[3:9, ]

age_group                                        review_text  \
clothing_id                                                                
2402              NaN                                                NaN   
2403              NaN                                                NaN   
767          25 to 34  Absolutely wonderful - silky and sexy and comf...   
1080         25 to 34  Love this dress!  it's sooo pretty.  i happene...   
1077         55 to 64  I had such high hopes for this dress and reall...   
1049         35 to 54  I love, love, love this jumpsuit. it's fun, fl...   

             rating  recommended_ind  positive_feedback_count   division_name  \
clothing_id                                                                     
2402            NaN              NaN                      NaN             NaN   
2403            NaN              NaN                      NaN             NaN   
767             4.0              1.0                      0.0       Initmates   
1080            5.0              1.0                      4.0         General   
1077            3.0              0.0                      0.0         General   
1049            5.0              1.0                      0.0  General Petite   

             category       review_date  
clothing_id                              
2402              NaN               NaN  
2403              NaN               NaN  
767          Intimate  2017-03-30 06:00  
1080          Dresses  2017-03-10 22:00  
1077          Dresses  2017-03-04 10:00  
1049          Bottoms               NaN

During the data exploration and preparation phase, it is likely we come across some problematic details in our data. This could be the value of _-1_ for the _age_ column, a value of _blank_ for the _customer segment_ column, or a value of _None_ for the _loan duration_ column. All of these are examples of "untidy" data, which is rather common depending on the data collection and recording process in a company.

In `pandas`, we use `NaN` (not a number) to denote missing data; The equivalent for datetime is `NaT` but both are essentially compatible with each other. From the docs:
> The choice of using `NaN` internally to denote missing data was largely for simplicity and performance reasons. We are hopeful that NumPy will soon be able to provide a native NA type solution (similar to R) performant enough to be used in pandas.

We can use `isna()` and `notna()` to detect missing values. An example code is as below:

In [117]:
review2['review_date'].isna()

clothing_id
2400     True
2401     True
910      True
2402     True
2403     True
767     False
1080    False
1077    False
1049     True
847     False
1080    False
858     False
858     False
1077    False
Name: review_date, dtype: bool

A common way of using the `.isna()` method is to combine it with the subsetting methods we've learned in previous lessons:

In [121]:
review2[review2['review_date'].isna()]

age_group                                        review_text  \
clothing_id                                                                
2400              NaN                                                NaN   
2401              NaN                                                NaN   
910               NaN                                                NaN   
2402              NaN                                                NaN   
2403              NaN                                                NaN   
1049         35 to 54  I love, love, love this jumpsuit. it's fun, fl...   

             rating  recommended_ind  positive_feedback_count   division_name  \
clothing_id                                                                     
2400            NaN              NaN                      NaN             NaN   
2401            NaN              NaN                      NaN             NaN   
910             NaN              NaN                      NaN             NaN   
2402            NaN              NaN                      NaN             NaN   
2403            NaN              NaN                      NaN             NaN   
1049            5.0              1.0                      0.0  General Petite   

            category review_date  
clothing_id                       
2400             NaN         NaN  
2401             NaN         NaN  
910              NaN         NaN  
2402             NaN         NaN  
2403             NaN         NaN  
1049         Bottoms         NaN

Go ahead and use `notna()` to extract all the rows where `weekday` column is not missing:

In [120]:
## Your code below


## -- Solution code

Another common use-case in missing values treatment is to count the number of `NAs` across each column:

In [124]:
review2.isna().sum()

age_group                  5
review_text                5
rating                     5
recommended_ind            5
positive_feedback_count    5
division_name              5
category                   5
review_date                6
dtype: int64

When we are certain that the rows with `NA`s can be safely dropped, we can use `dropna()`, optionally specifying a threshold. By default, this method drops the row if any NA value is present (`how='any'`), but it can be set to do this only when all values are NA in that row (`how='all'`).

```
    # drops row if all values are NA
    review2.dropna(how='all')
    
    # drops row if it doesn't have at least 5 non-NA values
    review2.dropna(thresh=5) 
```

In [126]:
review2.dropna(thresh=6).head()

age_group                                        review_text  \
clothing_id                                                                
767          25 to 34  Absolutely wonderful - silky and sexy and comf...   
1080         25 to 34  Love this dress!  it's sooo pretty.  i happene...   
1077         55 to 64  I had such high hopes for this dress and reall...   
1049         35 to 54  I love, love, love this jumpsuit. it's fun, fl...   
847          35 to 54  This shirt is very flattering to all due to th...   

             rating  recommended_ind  positive_feedback_count   division_name  \
clothing_id                                                                     
767             4.0              1.0                      0.0       Initmates   
1080            5.0              1.0                      4.0         General   
1077            3.0              0.0                      0.0         General   
1049            5.0              1.0                      0.0  General Petite   
847             5.0              1.0                      6.0         General   

             category       review_date  
clothing_id                              
767          Intimate  2017-03-30 06:00  
1080          Dresses  2017-03-10 22:00  
1077          Dresses  2017-03-04 10:00  
1049          Bottoms               NaN  
847              Tops  2017-12-14 06:00

When we have data where duplicated observations are recorded, we can use `.drop_duplicates()` specifying whether the first occurence or the last should be kept:

In [128]:
print(review2.shape)
print(review2.drop_duplicates(keep="first").shape)

(14, 8)
(10, 8)


### Knowledge Check
_Est. Time required: 20 minutes_

1. Duplicates may mean a different thing from a data point-of-view and a business analyst's point-of-view. You want to be extra careful about whether the duplicates is an intended characteristic of your data, or whether it poses a violation to the business logic. 

    - a. A medical center collects anonymized heart rate monitoring data from patients. It has duplicate observations collected across a span of 3 months
    - b. An insurance company uses machine learning to deliver dynamic pricing to its customers. Each row contains the customer's name, occupation / profession and historical health data. It has duplicate observations collected across a span of 3 months
    - c. On our original `review` data, check for duplicate observations. Would you have drop the duplicated rows?

---


In [139]:
review[review.duplicated(keep = False)].shape

(61, 8)


2. Once you've identified the missing values, there are 3 common ways to deal with it:

    - a. Use `dropna` with a reasonable threshold to remove any rows that contain too little values rendering it unhelpful to your analysis
    - b. Replace the missing values with a central value (mean or median)
    - c. Imputation through a predictive model
        - In a dataframe where `salary` is missing but the bank has data about the customer's occupation / profession, years of experience, years of education, seniority level, age, and industry, then a machine learning model such as regression or nearest neighbor can offer a viable alternative to the mean imputation approach
 
Going back to `review2`: what is a reasonable strategy? List them down or in pseudo-code.

In [140]:
## Your code below


## -- Solution code

#### Missing Values Treatment

Some common methods when working with missing values are demonstrated in the following section. We make a copy of the NA-included DataFrame, and name it `review3`:

In [143]:
review3 = review2.copy()
review3.head(7)

age_group                                        review_text  \
clothing_id                                                                
2400              NaN                                                NaN   
2401              NaN                                                NaN   
910               NaN                                                NaN   
2402              NaN                                                NaN   
2403              NaN                                                NaN   
767          25 to 34  Absolutely wonderful - silky and sexy and comf...   
1080         25 to 34  Love this dress!  it's sooo pretty.  i happene...   

             rating  recommended_ind  positive_feedback_count division_name  \
clothing_id                                                                   
2400            NaN              NaN                      NaN           NaN   
2401            NaN              NaN                      NaN           NaN   
910             NaN              NaN                      NaN           NaN   
2402            NaN              NaN                      NaN           NaN   
2403            NaN              NaN                      NaN           NaN   
767             4.0              1.0                      0.0     Initmates   
1080            5.0              1.0                      4.0       General   

             category       review_date  
clothing_id                              
2400              NaN               NaN  
2401              NaN               NaN  
910               NaN               NaN  
2402              NaN               NaN  
2403              NaN               NaN  
767          Intimate  2017-03-30 06:00  
1080          Dresses  2017-03-10 22:00

In the following cell, the technique is demonstrably repetitive or even verbose. This is done to give us an idea of all the different options we can pick from. 

You may observe, for example that the two lines of code are functionally identical:
- `.fillna(0)`
- `.replace(np.nan, 0)`

In [156]:
review3[['age_group','review_text', 'division_name', 'category']] = review3[['age_group','review_text', 'division_name', 'category']].fillna('Missing')
review3[['recommended_ind','positive_feedback_count']] = review3[['recommended_ind','positive_feedback_count']].fillna(0)
review3.review_date = review3.review_date.fillna(method='bfill')
review3.rating = review3.rating.fillna(review3.rating.median())

review3.head(10)

age_group                                        review_text  \
clothing_id                                                                
2400          Missing                                            Missing   
2401          Missing                                            Missing   
910           Missing                                            Missing   
2402          Missing                                            Missing   
2403          Missing                                            Missing   
767          25 to 34  Absolutely wonderful - silky and sexy and comf...   
1080         25 to 34  Love this dress!  it's sooo pretty.  i happene...   
1077         55 to 64  I had such high hopes for this dress and reall...   
1049         35 to 54  I love, love, love this jumpsuit. it's fun, fl...   
847          35 to 54  This shirt is very flattering to all due to th...   

             rating  recommended_ind  positive_feedback_count   division_name  \
clothing_id                                                                     
2400            5.0              0.0                      0.0         Missing   
2401            5.0              0.0                      0.0         Missing   
910             5.0              0.0                      0.0         Missing   
2402            5.0              0.0                      0.0         Missing   
2403            5.0              0.0                      0.0         Missing   
767             4.0              1.0                      0.0       Initmates   
1080            5.0              1.0                      4.0         General   
1077            3.0              0.0                      0.0         General   
1049            5.0              1.0                      0.0  General Petite   
847             5.0              1.0                      6.0         General   

             category       review_date  
clothing_id                              
2400          Missing           Missing  
2401          Missing           Missing  
910           Missing           Missing  
2402          Missing           Missing  
2403          Missing           Missing  
767          Intimate  2017-03-30 06:00  
1080          Dresses  2017-03-10 22:00  
1077          Dresses  2017-03-04 10:00  
1049          Bottoms           Missing  
847              Tops  2017-12-14 06:00

# Learn-by-Building
## Graded Assignment 2
### Exploratory Data Analysis

Read `amazon-electronic.csv` from the `data_input` folder. You may find it helpful to use `parse_dates=[__]` in the `read_csv()` call. This is the same data we worked on in the previous course.

Perform the necessary data preparation steps and use the exploratory data analysis techniques you've acquired to answer the questions below. 


1. Which brand is the favorite among TV & Video customers? 
2. If we convert our record to monthly period, when was the highest sales occur?
3. On that peak period, which merchant had the highest _sales volume_?
4. List top 5 best selling camera in 2017!
 